// @edgio/cli/commands/env/del.js
"use strict";

const EdgioPackageJson = require('../../utils/EdgioPackageJson');
const authenticate = require('../../prompts/authenticate');
const getEdgioConfig = require('../../utils/getEdgioConfig');
const getProperty = require('../../utils/getProperty');
const {
  EOL
} = require('os');
const {
  green
} = require('chalk');
const Validator = require('../../utils/validator');
const deprecatedCommandWarning = require('../../prompts/deprecatedCommandWarning');
exports.command = 'del';
// This command is hidden until console team will be ready to support it
// exports.describe = 'Deletes the environment with given name'
exports.builder = {
  environment: {
    type: 'string',
    alias: 'e',
    describe: 'Name of the environment.',
    demandOption: true
  },
  organization: {
    type: 'string',
    alias: 'o',
    describe: 'The name of the organization under which the property belongs. Using private space if omitted.'
  },
  property: {
    alias: 'p',
    type: 'string',
    describe: 'Slug of the property. Using package.json name property if omitted.'
  },
  path: {
    type: 'string',
    describe: "Path to your property's root directory. Uses current directory by default.",
    default: '.'
  },
  // @deprecated legacy params
  team: {
    type: 'string',
    alias: 't',
    deprecated: "Use '--organization' instead. Will be deprecated in next major version."
  },
  site: {
    type: 'string',
    alias: 's',
    deprecated: "Use '--property' instead. Will be deprecated in next major version."
  }
};
exports.handler = async yargs => {
  const {
    context,
    environment: environmentParam,
    site: siteSlugParam,
    property: propertySlugParam,
    team: teamSlugParam,
    organization: organizationSlugParam,
    path: pathParam
  } = yargs;
  await authenticate(context);
  if (siteSlugParam) {
    deprecatedCommandWarning(context, 'site', 'property');
  }
  if (teamSlugParam) {
    deprecatedCommandWarning(context, 'team', 'organization');
  }
  const config = getEdgioConfig();
  if (!config) {
    context.logger.error(`Error: edgio.config.js not found. Please run "0 init" to initialize your Edgio project.\n`);
    process.exit(1);
  }
  const validator = new Validator(context.logger, {
    exitOnError: true
  });
  const packageJson = EdgioPackageJson.loadPackageJson(pathParam);
  const organizationSlug = organizationSlugParam || teamSlugParam || config.organization || config.team;
  const propertySlug = validator.validatePropertyName(propertySlugParam || siteSlugParam || config.name || packageJson.name);
  context.logger.warn('WARNING: Experimental feature - compatibility and support is not guaranteed (APPOPS-15456).');
  await context.logger.title('⌛ Deleting environment...');
  const environmentDetails = [`> organization=${organizationSlug || 'Private space'}`, `> property=${propertySlug}`, `> environment=${environmentParam}`].filter(Boolean); // Removes falsy values
  context.logger.info(environmentDetails.join(EOL));
  let property = await getProperty(context, propertySlug, organizationSlug);
  let environments = property.environments.nodes;
  const environmentToDelete = environments.find(environment => environment.name === environmentParam);
  if (!environmentToDelete) {
    context.logger.error(`Error: Environment "${environmentParam}" was not found.`);
    process.exit(1);
  }
  const deleteResult = await context.api.deleteEnvironment(environmentToDelete.id);
  if (deleteResult.deleteEnvironment) {
    context.logger.info(green(`Environment "${environmentParam}" was successfully deleted.`));
    process.exit(0);
  }
  context.logger.error(`Error: Couldn't delete the environment "${environmentParam}"`);
  process.exit(1);
};
