// @edgio/cli/commands/env/list.js
"use strict";

const EdgioPackageJson = require('../../utils/EdgioPackageJson');
const authenticate = require('../../prompts/authenticate');
const getEdgioConfig = require('../../utils/getEdgioConfig');
const getProperty = require('../../utils/getProperty');
const {
  EOL
} = require('os');
const Validator = require('../../utils/validator');
const deprecatedCommandWarning = require('../../prompts/deprecatedCommandWarning');
exports.command = 'list';
// This command is hidden until console team will be ready to support it
// exports.describe = "Prints the list of property's environments"
exports.builder = {
  property: {
    alias: 'p',
    type: 'string',
    describe: 'Slug of the property. Using package.json name property by default.'
  },
  organization: {
    type: 'string',
    alias: 'o',
    describe: 'The name of the organization under which the property belongs. Using private space if omitted.'
  },
  path: {
    type: 'string',
    describe: "Path to your property's root directory. Uses current directory by default.",
    default: '.'
  },
  // @deprecated legacy params
  team: {
    type: 'string',
    alias: 't',
    deprecated: "Use '--organization' instead. Will be deprecated in next major version."
  },
  site: {
    type: 'string',
    alias: 's',
    deprecated: "Use '--property' instead. Will be deprecated in next major version."
  }
};
exports.handler = async yargs => {
  const {
    context,
    property: propertySlugParam,
    organization: organizationSlugParam,
    path: pathParam,
    // @deprecated legacy params
    site: siteSlugParam,
    team: teamSlugParam
  } = yargs;
  if (siteSlugParam) {
    deprecatedCommandWarning(context, 'site', 'property');
  }
  if (teamSlugParam) {
    deprecatedCommandWarning(context, 'team', 'organization');
  }
  const localPropertySlugParam = propertySlugParam || siteSlugParam;
  const localOrganizationSlugParam = organizationSlugParam || teamSlugParam;
  await authenticate(context);
  const config = getEdgioConfig();
  if (!config) {
    context.logger.error(`Error: edgio.config.js not found. Please run "0 init" to initialize your Edgio project.\n`);
    process.exit(1);
  }
  const validator = new Validator(context.logger, {
    exitOnError: true
  });
  const packageJson = EdgioPackageJson.loadPackageJson(pathParam);
  const propertySlug = validator.validatePropertyName(localPropertySlugParam || config.name || packageJson.name);
  const organizationSlug = localOrganizationSlugParam || config.organization;
  context.logger.warn('WARNING: Experimental feature - compatibility and support is not guaranteed (APPOPS-15456).');
  await context.logger.title('⌛ Loading environments...');
  const environmentDetails = [`> organization=${organizationSlug || 'Private space'}`, `> property=${propertySlug}${EOL}`].filter(Boolean); // Removes falsy values
  context.logger.info(environmentDetails.join(EOL));
  const property = await getProperty(context, propertySlug, organizationSlug);
  const environments = property.environments.nodes;
  await context.logger.title('Environments:');
  environments.forEach(environment => {
    context.logger.info(`${environment.name}`);
  });
};
