// @edgio/core/runtime/edge-functions/EdgeFunctionFetchError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFetchResponseErrorFriendlyMessage = void 0;
// Matches the definition of the `fetch_response_error` enum from mod_wasm.
const FetchResponseError = {
    UNKNOWN: 0,
    CONNECT_FAILED: 1,
    CONNECT_TIMEOUT: 2,
    SSL_CONNECT_FAILED: 3,
    SSL_CERTIFICATE_VALIDATION_FAILED: 4,
    BAD_REQUEST: 5,
    BAD_RESPONSE: 6,
    TIMEOUT: 7,
    REENTRANT_LOOP: 8,
    FORBIDDEN: 9,
};
// Matches the definition of the `fetch_response_error` enum from mod_wasm.
const FetchResponseErrorFriendlyMessage = {
    [FetchResponseError.UNKNOWN]: 'Unknown',
    [FetchResponseError.CONNECT_FAILED]: 'Connect Failed',
    [FetchResponseError.CONNECT_TIMEOUT]: 'Connect Timeout',
    [FetchResponseError.SSL_CONNECT_FAILED]: 'SSL Connect Failed',
    [FetchResponseError.SSL_CERTIFICATE_VALIDATION_FAILED]: 'SSL Certificate Validation Failed',
    [FetchResponseError.BAD_REQUEST]: 'Bad Request',
    [FetchResponseError.BAD_RESPONSE]: 'Bad Response',
    [FetchResponseError.TIMEOUT]: 'Timeout',
    [FetchResponseError.REENTRANT_LOOP]: 'Reentrant loop',
    [FetchResponseError.FORBIDDEN]: 'Forbidden',
};
function getFetchResponseErrorFriendlyMessage(errorCode) {
    return FetchResponseErrorFriendlyMessage[errorCode];
}
exports.getFetchResponseErrorFriendlyMessage = getFetchResponseErrorFriendlyMessage;
exports.default = FetchResponseError;
