// @edgio/core/runtime/edge-functions/EdgeFunctionsWasmManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This class is responsible for loading wasm files.
 */
class EdgeFunctionsWasmManager {
    /**
     * Returns the code of wasm file and ensures
     * that it is loaded only once.
     */
    static async getWasmFileFromCache(filename) {
        if (this.cache[filename])
            return this.cache[filename];
        this.cache[filename] = await this.getWasmFile(filename);
        return this.cache[filename];
    }
    /**
     * Returns the code of the QuickJS compiler
     */
    static async getQuickjsCompilerBytes() {
        return this.getWasmFileFromCache('quickjs-compiler.wasm');
    }
    /**
     * Returns the code of the QuickJS runtime
     */
    static async getQuickjsRuntimeBytes() {
        return this.getWasmFileFromCache('quickjs-runtime.wasm');
    }
    /**
     * Loads the WASM file from the filesystem or from the network depending on the used environment
     * and returns the content of the file as a Buffer.
     * When running in the browser the file is fetch from the root of the domain.
     * @param filename
     */
    static async getWasmFile(filename) {
        try {
            // Nodejs environment
            const { join } = require('path');
            const { promises } = require('fs');
            return promises.readFile(join(__dirname, filename));
        }
        catch {
            // Browser environment
            const res = await fetch(`/${filename}`);
            return res.arrayBuffer();
        }
    }
}
exports.default = EdgeFunctionsWasmManager;
EdgeFunctionsWasmManager.cache = {};
