// @edgio/core/runtime/edge-functions/EdgeFunctionWorkerParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/// This class is used to pass data from the main thread to the worker thread.
/// Everything in it must be serializable by v8 so that it can be serialized.
class EdgeFunctionWorkerParams {
    constructor(request, rawBody, path, sysVars, envVars, usrVars, httpVars, origins) {
        this.request = {
            httpVersion: request === null || request === void 0 ? void 0 : request.httpVersion,
            method: request === null || request === void 0 ? void 0 : request.method,
            rawBody,
            rawHeaders: request === null || request === void 0 ? void 0 : request.rawHeaders,
            url: `${request === null || request === void 0 ? void 0 : request.protocol}://${request === null || request === void 0 ? void 0 : request.headers['host']}${request === null || request === void 0 ? void 0 : request.url}`,
        };
        this.path = path;
        this.sysVars = sysVars;
        this.envVars = envVars;
        this.usrVars = usrVars;
        this.httpVars = httpVars;
        this.origins = origins;
    }
}
exports.default = EdgeFunctionWorkerParams;
