// @edgio/core/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaResponse = exports.LambdaRequest = exports.edgioRoutes = exports.exact = exports.not = exports.or = exports.and = exports.getConfig = exports.Router = void 0;
var Router_1 = require("./router/Router");
Object.defineProperty(exports, "Router", { enumerable: true, get: function () { return __importDefault(Router_1).default; } });
__exportStar(require("./constants"), exports);
var config_1 = require("./config");
Object.defineProperty(exports, "getConfig", { enumerable: true, get: function () { return config_1.getConfig; } });
var RouteCriteria_1 = require("./router/RouteCriteria");
Object.defineProperty(exports, "and", { enumerable: true, get: function () { return RouteCriteria_1.and; } });
Object.defineProperty(exports, "or", { enumerable: true, get: function () { return RouteCriteria_1.or; } });
Object.defineProperty(exports, "not", { enumerable: true, get: function () { return RouteCriteria_1.not; } });
var exact_1 = require("./router/exact");
Object.defineProperty(exports, "exact", { enumerable: true, get: function () { return __importDefault(exact_1).default; } });
var router_1 = require("./router");
Object.defineProperty(exports, "edgioRoutes", { enumerable: true, get: function () { return router_1.edgioRoutes; } });
var LambdaRequest_1 = require("./runtime/LambdaRequest");
Object.defineProperty(exports, "LambdaRequest", { enumerable: true, get: function () { return __importDefault(LambdaRequest_1).default; } });
var LambdaResponse_1 = require("./runtime/LambdaResponse");
Object.defineProperty(exports, "LambdaResponse", { enumerable: true, get: function () { return __importDefault(LambdaResponse_1).default; } });
