// @edgio/core/origins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPrefetchOrigin = exports.createImageOptimizerOrigin = exports.createServerlessOrigin = exports.createPermanentStaticOrigin = exports.createStaticOrigin = exports.getPathPrefix = exports.getEdgioOrigins = exports.PREFETCH_ORIGIN_NAME = exports.IMAGE_OPTIMIZER_ORIGIN_NAME = exports.SERVERLESS_ORIGIN_NAME = exports.PERMANENT_STATIC_ORIGIN_NAME = exports.STATIC_ORIGIN_NAME = void 0;
const constants_1 = require("./constants");
const environment_1 = require("./environment");
/**
 * The id of the origin from which we serve static assets associated with a single deployment
 */
exports.STATIC_ORIGIN_NAME = 'edgio_static';
/**
 * The id of the origin from which we serve static assets with globally unique filenames (usually based on a hash of the file content)
 * for all deployments.
 */
exports.PERMANENT_STATIC_ORIGIN_NAME = 'edgio_permanent_static';
/**
 * The id of the origin from which serverless functions are served.
 */
exports.SERVERLESS_ORIGIN_NAME = 'edgio_serverless';
/**
 * The id of the origin from which we serve our image optimizer locally.
 * When app is deployed to Edgio, the SERVERLESS origin is used instead.
 */
exports.IMAGE_OPTIMIZER_ORIGIN_NAME = 'edgio_image_optimizer';
/**
 * The id of the origin with pre-built @edgio/prefetch assets
 */
exports.PREFETCH_ORIGIN_NAME = 'edgio_prefetch';
const internalConfig = process.env[constants_1.EDGIO_ENV_VARIABLES.internalConfig]
    ? JSON.parse(process.env[constants_1.EDGIO_ENV_VARIABLES.internalConfig] || '{}')
    : {};
// TODO: we need this function to remove path_prefix from the origin config
// this is currently added so we can do propper url rewriting in the edge,
// but once this is not needed we should remove it.
// Without this function, Sailfish validation will fail
const getOriginFromConfig = (name) => {
    var _a, _b;
    const origin = (_a = internalConfig === null || internalConfig === void 0 ? void 0 : internalConfig.origins) === null || _a === void 0 ? void 0 : _a.find((origin) => origin.name === name);
    if (origin) {
        for (let host of origin.hosts) {
            // @ts-ignore
            delete host.path_prefix;
        }
        return {
            ...origin,
            // This is temporary fix until this bug is fixed in sailfish: POR-31143
            // The Sailfish is returning 502 error for all other origins without use_sni
            // when first origin has tls_verify.use_sni set to true. The serverless origin is not working because of it.
            // TODO: Remove it when the POR-31143 is fixed.
            tls_verify: {
                use_sni: true,
                sni_hint_and_strict_san_check: (_b = origin === null || origin === void 0 ? void 0 : origin.hosts.find((host) => host === null || host === void 0 ? void 0 : host.location)) === null || _b === void 0 ? void 0 : _b.location,
            },
        };
    }
    return origin;
};
function getEdgioOrigins(forEdgeControl = false) {
    return [
        createStaticOrigin(),
        createPermanentStaticOrigin(),
        createServerlessOrigin(forEdgeControl),
        createImageOptimizerOrigin(),
        // edgio_prefetch origin is added just for local simulation
        ...((0, environment_1.isLocal)() ? [createPrefetchOrigin()] : []),
    ];
}
exports.getEdgioOrigins = getEdgioOrigins;
function getPathPrefix(origin) {
    var _a, _b, _c;
    return (((_c = (_b = (_a = internalConfig === null || internalConfig === void 0 ? void 0 : internalConfig.origins) === null || _a === void 0 ? void 0 : _a.find((o) => o.name === origin)) === null || _b === void 0 ? void 0 : _b.hosts[0]) === null || _c === void 0 ? void 0 : _c.path_prefix) || '');
}
exports.getPathPrefix = getPathPrefix;
/**
 *  * Creates the origin group for serving static assets associated with a single deployment.
 * @returns
 */
function createStaticOrigin() {
    return (getOriginFromConfig(exports.STATIC_ORIGIN_NAME) || {
        name: exports.STATIC_ORIGIN_NAME,
        hosts: [{ location: '127.0.0.1:3002' }],
    });
}
exports.createStaticOrigin = createStaticOrigin;
/**
 * Creates the origin group for serving static assets with globally unique URLs across all deployments.
 * @returns
 */
function createPermanentStaticOrigin() {
    return (getOriginFromConfig(exports.PERMANENT_STATIC_ORIGIN_NAME) || {
        name: exports.PERMANENT_STATIC_ORIGIN_NAME,
        hosts: [{ location: '127.0.0.1:3002' }],
    });
}
exports.createPermanentStaticOrigin = createPermanentStaticOrigin;
/**
 * Creates the origin group for serverless functions.
 * When forEdgeControl is false, the serverless origin target is the app. (In simulator and on permalink)
 * When forEdgeControl is true, the serverless origin target is link to permalink.
 * This ensure that the app is not infinitely sending requests to itself.
 * @returns
 */
function createServerlessOrigin(forEdgeControl = false) {
    // Return origin from config only in case we are generating config for edge control.
    // Otherwise keep the localhost origin which is pointing to the app.
    const originFromConfig = forEdgeControl ? getOriginFromConfig(exports.SERVERLESS_ORIGIN_NAME) : null;
    return (originFromConfig || {
        name: exports.SERVERLESS_ORIGIN_NAME,
        hosts: [{ location: '127.0.0.1:3001' }],
    });
}
exports.createServerlessOrigin = createServerlessOrigin;
/**
 * Creates the origin for the image optimizer.
 * This origin is used when we run the app locally in production mode.
 * When app is deployed to Edgio, the SERVERLESS origin is used instead.
 * @returns
 */
function createImageOptimizerOrigin() {
    var _a;
    const serverlessOriginConfig = getOriginFromConfig(exports.SERVERLESS_ORIGIN_NAME);
    const optimizerOriginConfig = (0, environment_1.isLocal)()
        ? {
            hosts: [{ location: '127.0.0.1:3003' }],
        }
        : {
            ...serverlessOriginConfig,
            override_host_header: (_a = serverlessOriginConfig === null || serverlessOriginConfig === void 0 ? void 0 : serverlessOriginConfig.hosts[0]) === null || _a === void 0 ? void 0 : _a.location,
        };
    return {
        ...optimizerOriginConfig,
        name: exports.IMAGE_OPTIMIZER_ORIGIN_NAME,
    };
}
exports.createImageOptimizerOrigin = createImageOptimizerOrigin;
/**
 * Creates the origin config for the prefetch origin with pre-built @edgio/prefetch assets.
 * This origin is used just for local simulation.
 * @returns
 */
function createPrefetchOrigin() {
    return {
        name: exports.PREFETCH_ORIGIN_NAME,
        override_host_header: 'edgio-sites-prefetch-default.edgio.link',
        hosts: [{ location: 'edgio-sites-prefetch-default.edgio.link' }],
    };
}
exports.createPrefetchOrigin = createPrefetchOrigin;
