// @edgio/cli/frameworks/frameworks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
// istanbul ignore file
// Note: order of keys are important. If we have more
// specific version of the same key, then specific version
// of the same key must go before the less specific version
// for example:
// if we have:
//
// {
//    key: "nuxt",
//    name: "Nuxt.js"
//    builder: "@edgio\nuxt-nitro",
//    framework: 'nuxt',
//    frameworkVersion: "3.x"  <-- specific version must be placed before non specific
//    ...
// }
// {
//    key: "nuxt",
//    name: "Nuxt.js"
//    builder: "@edgio\nuxt",
//    framework: 'nuxt',
//    ...
// }
//
// The versions are compared using "satisfies" function from https://www.npmjs.com/package/semver
//
var _default = [{
  key: 'next',
  name: 'Next.js',
  builder: '@edgio/next',
  devDependencies: ['@edgio/react', '@edgio/devtools'],
  framework: 'next',
  frameworkVersion: undefined
}, {
  key: 'angular',
  name: 'Angular',
  builder: '@edgio/connectors',
  devDependencies: ['globby@11.0.2'],
  // later Angular versions use globby@12.x which breaks build
  framework: '@angular/core',
  frameworkVersion: undefined
}, {
  key: 'nuxt',
  name: 'Nuxt',
  builder: '@edgio/nuxt-nitro',
  devDependencies: ['@edgio/vue', '@edgio/devtools'],
  framework: 'nuxt',
  frameworkVersion: '>=3.0.0'
}, {
  key: 'nuxt-v2',
  name: 'Nuxt v2',
  builder: '@edgio/nuxt',
  devDependencies: ['@edgio/vue'],
  framework: 'nuxt',
  frameworkVersion: undefined
}, {
  key: 'vue-cva',
  name: 'Vue',
  builder: '@edgio/connectors',
  devDependencies: ['@edgio/vue'],
  framework: 'vue',
  frameworkVersion: '>=2.0.0'
}, {
  key: 'sveltekit',
  name: 'SvelteKit',
  builder: '@edgio/connectors',
  dependencies: ['@edgio/sveltekit'],
  devDependencies: ['@edgio/svelte'],
  framework: '@sveltejs/kit',
  frameworkVersion: undefined
}, {
  key: 'astro',
  name: 'astro',
  builder: '@edgio/connectors',
  framework: 'astro',
  frameworkVersion: '>=1.0.0'
}, {
  key: 'qwik',
  name: 'qwik',
  builder: '@edgio/connectors',
  framework: '@builder.io/qwik'
}, {
  key: 'react-cra',
  name: 'react-cra',
  builder: '@edgio/connectors',
  devDependencies: ['@edgio/devtools'],
  framework: 'react-scripts'
},
// express should be at the end, as express could be used by other frameworks
// as a server, for instance, qwik can use express as one of the serving server
{
  key: 'express',
  name: 'express',
  builder: '@edgio/connectors',
  framework: 'express',
  frameworkVersion: undefined
}];
exports.default = _default;
module.exports = exports.default;
