// @edgio/core/runtime/createEdgeConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEdgeConfigPropertiesObject = void 0;
const RulesFormatter_1 = __importDefault(require("./rules/RulesFormatter"));
const edge_functions_1 = require("../deploy/edge-functions");
/**
 * Creates the JSON for deploying an environment via the EdgeControl API.
 * @param context The PropertyContext
 * @param router The Router use to create rules
 * @param isEdge set to true if not running in simulator
 * @returns
 */
function createEdgeConfig(context, router) {
    return JSON.stringify(createEdgeConfigPropertiesObject(context, router), null, 2);
}
exports.default = createEdgeConfig;
/**
 * Create edge control json object
 * @param context The PropertyContext
 * @param router The Router use to create rules
 * @returns
 */
function createEdgeConfigPropertiesObject(context, router) {
    const sources = (0, edge_functions_1.getEdgeFunctionSources)();
    const includeEdgeFunctions = Object.keys(sources).length > 0;
    const edgeConfig = {
        name: context.property.name,
        hostnames: context.property.hostnames,
        origins: context.origins,
        rules: router.rules,
        // Don't include edge functions if none are defined.
        edge_functions: includeEdgeFunctions
            ? {
                ...context.property.edge_functions,
                sources: sources,
            }
            : undefined,
    };
    // Adjust rules when we are exporting config for EdgeControl.
    const rulesFormatter = new RulesFormatter_1.default();
    const rules = rulesFormatter.format(router.rules);
    return {
        ...edgeConfig,
        rules,
    };
}
exports.createEdgeConfigPropertiesObject = createEdgeConfigPropertiesObject;
