// @edgio/core/runtime/PropertyContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class PropertyContext {
    constructor(property) {
        var _a;
        this.accountID = process.env.ACCOUNT_ID || 'development';
        this.property = property;
        this.origins = (_a = this.property.origins) !== null && _a !== void 0 ? _a : [];
        this.originsByName = this.createOriginMap();
        this.defaultOrigins = this.createDefaultOriginMap();
    }
    /**
     * Gets an origin by name
     * @param name
     * @returns
     */
    getOrigin(name) {
        return this.originsByName[name];
    }
    /**
     * Return default origin name based on the
     * matching the host header to the corresponding cname's default origin from edgio.config.js.
     * @param request
     */
    getDefaultOrigin(request) {
        let hostname = request.getHeader('host').toLowerCase();
        const defaultOrigin = this.defaultOrigins[hostname];
        const origins = this.origins;
        if (defaultOrigin && defaultOrigin.name) {
            return defaultOrigin.name;
        }
        return origins[0].name;
    }
    /**
     * Creates a map of origin group id to origin group
     * @returns
     */
    createOriginMap() {
        const originGroups = {};
        const origins = this.origins;
        for (const origin of origins) {
            originGroups[origin.name] = origin;
        }
        return originGroups;
    }
    /**
     * Creates a map of hostname to the default origin group
     * @returns
     */
    createDefaultOriginMap() {
        const defaultOriginMap = {};
        const { hostnames } = this.property;
        if (hostnames) {
            for (const hostname of hostnames) {
                if (hostname.default_origin_name) {
                    const origin = this.originsByName[hostname.default_origin_name];
                    if (origin) {
                        defaultOriginMap[hostname.hostname] = origin;
                    }
                    else {
                        throw new Error(`No origin with id="${hostname.default_origin_name}" was found.`);
                    }
                }
            }
        }
        return defaultOriginMap;
    }
}
exports.default = PropertyContext;
