// @edgio/core/router/RouteCriteria.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.not = exports.or = exports.and = exports.OrOperator = exports.AndOperator = exports.isNegation = void 0;
const isNegation = (value) => {
    return !!value && typeof value === 'object' && 'not' in value;
};
exports.isNegation = isNegation;
class CriteriaOperator {
    constructor(values) {
        this.values = values;
    }
}
class AndOperator extends CriteriaOperator {
    constructor() {
        super(...arguments);
        this.operator = 'and';
    }
}
exports.AndOperator = AndOperator;
class OrOperator extends CriteriaOperator {
    constructor() {
        super(...arguments);
        this.operator = 'or';
    }
}
exports.OrOperator = OrOperator;
/**
 * Creates an AND operator for router's condition functions
 */
function and(...criterias) {
    return new AndOperator(criterias);
}
exports.and = and;
/**
 * Creates an OR operator for router's condition functions
 */
function or(...criterias) {
    return new OrOperator(criterias);
}
exports.or = or;
/**
 * Creates negation for router's condition
 */
function not(value) {
    return {
        not: value,
    };
}
exports.not = not;
