// @edgio/prefetch/sw/Prefetcher.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stats = void 0;
const workbox_routing_1 = require("workbox-routing");
const workbox_strategies_1 = require("workbox-strategies");
const CustomExpirationPlugin_1 = require("./CustomExpirationPlugin");
const workbox_cacheable_response_1 = require("workbox-cacheable-response");
const constants_1 = require("../constants");
const log_1 = __importDefault(require("./log"));
const PrefetchPlugin_1 = __importDefault(require("./PrefetchPlugin"));
const DevtoolsPlugin_1 = __importDefault(require("./DevtoolsPlugin"));
const messageBrowser_1 = __importDefault(require("./messageBrowser"));
const constants_2 = require("../constants");
class Stats {
    constructor() {
        this.hits = 0;
        this.misses = 0;
    }
}
exports.Stats = Stats;
const BUFFER_DELAY_MS = 10000;
/**
 * Handles prefetching and caching in the service worker. Responses
 * are cached based on the `service_worker_max_age` feature
 * or equivalent `browser.serviceWorkerSeconds` cache setting in your rules.
 *
 * **Example routes.js**
 *
 * ```js
 *  import { Router } from '@edgio/core/router'
 *
 *  module.exports = new Router()
 *    .match('/some/path', {
 *       caching: {
 *          max_age: "1d",
 *          service_worker_max_age: "1h"
 *       }
 *     })
 *   .match('/another/path', ({ cache }) => {
 *      cache({
 *        edge: {
 *          maxAgeSeconds: 60 * 60 * 24 // one day
 *        },
 *        browser: {
 *          serviceWorkerSeconds: 60 * 60 // one hour
 *        }
 *      })
 *    })
 * ```
 *
 * **Example Service Worker**
 *
 * ```js
 *  import { skipWaiting, clientsClaim } from 'workbox-core'
 *  import { Prefetcher } from '@edgio/prefetch/sw'
 *  import { precacheAndRoute } from 'workbox-precaching'
 *
 *  skipWaiting()
 *  clientsClaim()
 *  precacheAndRoute(self.__WB_MANIFEST || [])
 *
 *  new Prefetcher().route()
 * ```
 */
class Prefetcher {
    constructor(config = { plugins: [] }) {
        this.skipAssets = false;
        // For performance reasons, Devtools plugins are disabled until the first
        // 'edgio-devtools-syn' message is received (ie: a Devtools client is enabled).
        // But once enabled, they will remain active until the Service Worker is reinstalled
        this.devtoolsEnabled = false;
        /**
         * bufferedDevtoolMessages keeps track of the request received in the last BUFFER_DELAY_MS.
         */
        this.bufferedDevtoolMessages = [];
        this.config = config;
        this.cacheOrigins = [self.origin];
        if (config.cacheHost) {
            this.cacheHost = new URL(`https://${config.cacheHost}`);
            this.cacheOrigins.push(this.cacheHost.origin);
        }
        this.cacheableResponsePlugin = new workbox_cacheable_response_1.CacheableResponsePlugin({
            statuses: [200, 301, 302],
        });
        this.skipAssets = config.skipAssets || false;
        this.stats = new Stats();
        this.initStats();
        this.initDevtoolsHandshakeListener();
    }
    /**
     * Creates a fallback route that ensures that all requests are handled by the service worker so that
     * we properly cache (or not cache) based on x-sw-cache-control, and that all network requests are seen by the
     * Devtools.
     */
    createFallbackRoute() {
        this.createRoute({
            pattern: '^.*$',
            maxAgeSeconds: 0,
            isPartialPathRoute: true,
        });
    }
    /**
     * Adds a fetch event listener that serves content from the browser cache when available.  When
     * a request is not in the cache, it will be fetched from the network and added to the
     * cache if the route has `service_worker_max_age' feature defined or if the request is a prefetch.
     */
    route() {
        this.createFallbackRoute();
        return this;
    }
    cache(pattern, maxAgeSeconds = 60 * 60 * 24) {
        this.createRoute({ pattern: pattern.source, maxAgeSeconds, isPartialPathRoute: false });
        return this;
    }
    /**
     * Starts listening for messages from the browser
     */
    initStats() {
        self.addEventListener('message', (e) => {
            if (e.data.action === 'get-stats') {
                (0, messageBrowser_1.default)({
                    type: 'stats',
                    stats: this.stats,
                });
            }
        });
    }
    /**
     * Here we implement a handshake protocol between the devtool window client and the service worker:
     * 1. Devtools sends the 'edgio-devtools-syn' event and starts listening for 'edgio-devtools-ack' event
     * 2. Service worker receives the 'edgio-devtools-syn' event and sends the 'edgio-devtools-ack' event
     *    along with the buffered messages that were received over the last BUFFER_DELAY_MS.
     *    This is because the window widget takes time to download, load and to start listening from
     *    Service worker message, while the latter is already receiving http requests.
     * 3. Devtools receives the 'edgio-devtools-ack' event, looks for its current page html request,
     *    which corresponds to that tab very first request, and loads requests past that one.
     * 4. Devtools starts listening for live 'edgio-devtools-request' events
     *
     * Notes:
     * - there is no 1-1 channel between window client and Service Worker, so all opened tab
     *   will see the 'edgio-devtools-ack' event. Only the ones in handshake phase will handle it.
     */
    initDevtoolsHandshakeListener() {
        self.addEventListener('message', async (e) => {
            if (e.data.action === 'edgio-devtools-install') {
                if (!this.devtoolsEnabled) {
                    this.devtoolsEnabled = true;
                    // Devtools plugin just got initialized so we missed
                    // a few queries. We send a message so that the browser
                    // refreshes its page.
                    this.sendToAllClients({
                        type: 'edgio-devtools-ack-installed',
                    });
                }
            }
            if (e.data.action === 'edgio-devtools-syn') {
                if (this.devtoolsEnabled) {
                    const ackMessage = {
                        type: 'edgio-devtools-ack',
                        bufferedDevtoolMessages: this.bufferedDevtoolMessages,
                    };
                    this.sendToAllClients(ackMessage);
                }
            }
            if (e.data.action === 'edgio-devtools-clear-cache') {
                caches.keys().then(cacheNames => cacheNames.forEach(name => caches.delete(name)));
            }
        });
    }
    /**
     * Send a postMessage event to all window client.
     * Notes:
     * - When used in multi-tabs, each Devtools will show data from all tabs.
     *   That's a bit buggy but it's the best we can do as it's very hard to establish 1-1 channels
     *   on ServiceWorker/Client interface, for example the message event from the window does not
     *   contain the clientId
     */
    async sendToAllClients(data) {
        (0, messageBrowser_1.default)(data, { broadcast: true });
    }
    /**
     * Adds a message to this.bufferedDevtoolMessages and manages the buffer so that
     * we don't keep messages more than BUFFER_DELAY_MS.
     *
     * In case of inactivity we clear the whole buffer after BUFFER_DELAY_MS, so technically
     * a message can stay up to 2 x BUFFER_DELAY_MS in the buffer (a bit less than BUFFER_DELAY_MS old
     * when the latest message arrives and cleared BUFFER_DELAY_MS later)
     * @param message
     */
    addDevtoolsRequestMessageToBuffer(message) {
        this.bufferedDevtoolMessages.push(message);
        this.trimDevtoolsRequestMessageBuffer();
        if (this.clearDevtoolsRequestMessageTimeout) {
            self.clearTimeout(this.clearDevtoolsRequestMessageTimeout);
        }
        // If no additional request for BUFFER_DELAY_MS, we clear the buffer after that
        this.clearDevtoolsRequestMessageTimeout = self.setTimeout(() => {
            this.bufferedDevtoolMessages = [];
        }, BUFFER_DELAY_MS);
    }
    /**
     * Remove message that are too old to stay in devtools message buffer
     */
    trimDevtoolsRequestMessageBuffer() {
        let sliceIndex = 0;
        const sliceMinimumTimestamp = Date.now() - BUFFER_DELAY_MS;
        while (this.bufferedDevtoolMessages[sliceIndex] &&
            this.bufferedDevtoolMessages[sliceIndex].timestamp < sliceMinimumTimestamp) {
            sliceIndex++;
        }
        this.bufferedDevtoolMessages = this.bufferedDevtoolMessages.slice(sliceIndex);
    }
    /**
     * If there's a cacheHost configured check whether the pattern to test
     * against is a partial pathname pattern. If so, check that the origin is
     * also correct: service worker's own origin or cacheHost origin
     * @param pattern A route pattern
     * @param isPartialPathRoute Set to true if the supplied pattern only matches the path, false if it matches the whole URL.
     * @return true if the request should be handled by the service worker.
     */
    shouldIntercept(pattern, isPartialPathRoute) {
        return ({ url, request }) => {
            if (this.skipAssets &&
                request.destination &&
                !['document', 'empty'].includes(request.destination)) {
                return false;
            }
            if (isPartialPathRoute) {
                return this.cacheOrigins.includes(url.origin) && pattern.test(url.pathname);
            }
            else {
                return pattern.test(url.toString());
            }
        };
    }
    /**
     * Creates a DevtoolsPlugin object to be added to a caching strategy instance
     * @param routeConfig The route config for the given route
     * @private
     */
    createDevtoolsPlugin(routeConfig) {
        return new DevtoolsPlugin_1.default({
            onResponse: this.onResponse.bind(this),
            isEnabledFn: () => this.devtoolsEnabled,
            routeConfig,
        });
    }
    /**
     * Creates a workbox route.
     * @param routeConfig The route pattern and caching options specified within the route definition
     */
    createRoute(routeConfig) {
        const { pattern, method: optMethod, maxAgeSeconds, isPartialPathRoute, convertToGet, prefetchUpstreamRequests, } = routeConfig;
        const method = (optMethod === null || optMethod === void 0 ? void 0 : optMethod.value) || 'get';
        const patternRegexp = new RegExp(pattern, 'i');
        (0, log_1.default)(`[route] ${method} ${pattern}, maxAgeSeconds: ${maxAgeSeconds}`);
        // all routes must have a GET route registered, so Workbox can properly look
        // it up when running workbox-routing/Router#findMatchingRoute:
        if (method !== 'get') {
            this.createRoute({ ...routeConfig, method: { value: 'get' } });
        }
        const prefetchPlugin = new PrefetchPlugin_1.default({
            incrementRead: this.incrementRead.bind(this),
            incrementWrite: this.incrementWrite.bind(this),
            cacheHost: this.cacheHost,
            prefetchUpstreamRequests,
            convertToGet,
            method,
        });
        const ttl = maxAgeSeconds || this.config.defaultMaxAgeSeconds || constants_2.DEFAULT_MAX_AGE_SECONDS;
        const plugins = [
            this.cacheableResponsePlugin,
            new CustomExpirationPlugin_1.CustomExpirationPlugin({ maxAgeSeconds: ttl }),
            // DevtoolPlugin needs to be injected before PrefetchPlugin as this later
            // changes request url to normalise the cache key
            this.createDevtoolsPlugin(routeConfig),
            prefetchPlugin,
            ...(this.config.plugins || []),
        ];
        (0, workbox_routing_1.registerRoute)(this.shouldIntercept(patternRegexp, isPartialPathRoute), new workbox_strategies_1.CacheFirst({
            cacheName: constants_1.CACHE_NAME,
            matchOptions: {
                ignoreVary: true, // without this, prefetched responses will not be served from the cache during navigation
            },
            plugins,
        }), this.transformHTTPMethod(method) // HTTPMethods map directly from @edgio/core lowercase => @workbox uppercase
        );
    }
    /**
     * Transform method from HTTPMethod type to uppercase HTTPMethod type from workbox
     * the way typescript tolerates
     * @param method
     * @returns
     */
    transformHTTPMethod(method) {
        if (method == 'delete')
            return 'DELETE';
        if (method == 'get')
            return 'GET';
        if (method == 'head')
            return 'HEAD';
        if (method == 'patch')
            return 'PATCH';
        if (method == 'post')
            return 'POST';
        if (method == 'put')
            return 'PUT';
        return undefined;
    }
    /**
     * A Read is considered a Hit until a Write happens meaning it was a miss
     * This is based on the fragile assumption that a cache read *always* triggers a write
     * in case of cache miss, which is not true if the backend does not respond for example.
     *
     * This could be fixed by checking the presence of `response` in `cachedResponseWillBeUsed``
     * but that leads to other issues as we cannot discard Prefetch requests at that
     * stage.
     * We could implement the same Hack-ish solution as in DevtoolsPlugin to work around that,
     * but those simple stats are probably going away anyway.
     */
    incrementRead() {
        this.stats.hits++;
    }
    incrementWrite() {
        this.stats.misses++;
        this.stats.hits--;
    }
    async onResponse({ request, response, ttfb, prefetch, browserCache, routeConfig, }) {
        const headers = {};
        response.headers.forEach((value, key) => {
            headers[key] = value;
        });
        const devToolMessage = {
            type: 'edgio-devtools-request',
            timestamp: Date.now(),
            request: {
                method: request.method,
                url: request.url,
                prefetch,
            },
            response: {
                ttfb,
                status: response.status,
                headers,
            },
            browserCache,
            routeConfig,
        };
        this.sendToAllClients(devToolMessage);
        this.addDevtoolsRequestMessageToBuffer(devToolMessage);
    }
}
exports.default = Prefetcher;
